pkg/apis/wksprovider/controller/wksctl/cluster_controller.go

```diff
@@ -0,0 +1,100 @@
+package wks
+
+import (
+	"context"
+	"errors"
+
+	log "github.com/sirupsen/logrus"
+	baremetalspecv1 "github.com/weaveworks/wksctl/pkg/baremetal/v1alpha3"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+// ClusterReconciler is responsible for managing this cluster, and ensuring its
+// state converges towards its definition.
+type ClusterReconciler struct {
+	client        client.Client
+	eventRecorder record.EventRecorder
+}
+
+// Reconcile creates or updates the cluster.
+func (a *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.TODO() // upstream will add this eventually
+	contextLog := log.WithField("name", req.NamespacedName)
+
+	// request only contains the name of the object, so fetch it from the api-server
+	bmc := &baremetalspecv1.BareMetalCluster{}
+	err := a.client.Get(ctx, req.NamespacedName, bmc)
+	if err != nil {
+		if apierrs.IsNotFound(err) { // isn't there; give in
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Get Cluster via OwnerReferences
+	cluster, err := util.GetOwnerCluster(ctx, a.client, bmc.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		contextLog.Info("Cluster Controller has not yet set ownerReferences")
+		return ctrl.Result{}, nil
+	}
+	contextLog = contextLog.WithField("cluster", cluster.Name)
+
+	if util.IsPaused(cluster, bmc) {
+		contextLog.Info("BareMetalCluster or linked Cluster is marked as paused. Won't reconcile")
+		return ctrl.Result{}, nil
+	}
+
+	// Object still there but with deletion timestamp => run our finalizer
+	if !bmc.ObjectMeta.DeletionTimestamp.IsZero() {
+		a.recordEvent(cluster, corev1.EventTypeNormal, "Delete", "Deleted cluster %v", cluster.Name)
+		return ctrl.Result{}, errors.New("ClusterReconciler#Delete not implemented")
+	}
+
+	// FIXME!  do we need any more?
```

Comment:
The BareMetalCluster controller is essentially a noop controller. If we implement `status` then we sould try to set the status properties here.